docs/autogen.py

```diff
@@ -268,204 +51,31 @@ def add_np_implementation(function, docstring):
-            subblocks.append(code_snippet(signature))
-            docstring = cls.__doc__
-            if docstring:
-                subblocks.append(process_docstring(docstring))
-            methods = collect_class_methods(cls, element[1])
-            if methods:
-                subblocks.append('\n---')
-                subblocks.append('## ' + cls.__name__ + ' methods\n')
-                subblocks.append('\n---\n'.join(
-                    [render_function(method, method=True)
-                     for method in methods]))
-            blocks.append('\n'.join(subblocks))
-
-        methods = read_page_data(page_data, 'methods')
-
-        for method in methods:
-            blocks.append(render_function(method, method=True))
-
-        functions = read_page_data(page_data, 'functions')
-
-        for function in functions:
-            blocks.append(render_function(function, method=False))
-
-        if not blocks:
-            raise RuntimeError('Found no content for page ' +
-                               page_data['page'])
+def preprocess_docstring(docstring, function, signature):
+    if "backend" in signature and "{{np_implementation}}" in docstring:
+        docstring = add_np_implementation(function, docstring)
+    return docstring
 
-        mkdown = '\n----\n\n'.join(blocks)
-        # Save module page.
-        # Either insert content into existing page,
-        # or create page otherwise.
-        page_name = page_data['page']
-        path = os.path.join(sources_dir, page_name)
-        if os.path.exists(path):
-            template = read_file(path)
-            if '{{autogenerated}}' not in template:
-                raise RuntimeError('Template found for ' + path +
-                                   ' but missing {{autogenerated}}'
-                                   ' tag.')
-            mkdown = template.replace('{{autogenerated}}', mkdown)
-            print('...inserting autogenerated content into template:', path)
-        else:
-            print('...creating new page with autogenerated content:', path)
-        subdir = os.path.dirname(path)
-        if not os.path.exists(subdir):
-            os.makedirs(subdir)
-        with open(path, 'w', encoding='utf-8') as f:
-            f.write(mkdown)
 
-    shutil.copyfile(os.path.join(str(keras_dir), 'CONTRIBUTING.md'),
-                    os.path.join(str(sources_dir), 'contributing.md'))
-    copy_examples(os.path.join(str(keras_dir), 'examples'),
-                  os.path.join(str(sources_dir), 'examples'))
+def generate(dest_dir):
+    template_dir = keras_dir / "docs" / "templates"
+    keras_autodoc.generate(
+        dest_dir,
+        template_dir,
+        PAGES,
+        'https://github.com/keras-team/keras/blob/master',
+        keras_dir / "examples",
+        EXCLUDE,
+        clean_module_name=clean_module_name,
+        post_process_signature=post_process_signature,
+        preprocess_docstring=preprocess_docstring,
```

Comment:
This API appears to be shaped by the ad-hoc constraints of Keras docs generation. It order to make it more abstract and configurable, we could:

- Expose `process_signature()`, `process_docstring()` name as public utility functions
- Enable the use to pass overriding arguments for them (the user-defined function would themselves make use of the default `process_signature()` and `process_docstring()` methods)

This gives the user full control under a simple API.

Note that a custom `clean_module_name` can simply be part of `process_signature`, it doesn't need to be special-cased.

Reply:
I really didn't give much thoughts to the API. Can we merge this PR and I'll do the API change you propose in another PR? It'll be easier for me to work with a smaller git diff.

Reply:
I think it would be simpler to start by refactoring autodocs into a generic API, and then make the PR to make Keras rely on it. Autodoc's ability to support both Keras, AutoKeras and Keras Tuner with a single API would be a good test of how generic it is.

Reply:
Sure thing. Give me some time.

Reply:
I'm working on some refactoring and improving readability. It takes a bit of time.

Reply:
Done. I've cleared up a lot of things and even found some bugs that I fixed. I favored a class-subclass approach as the way to override default functions. Now the API is centralized around a single class:

```python

class DocumentationGenerator:

    def __init__(self,
                 pages=None,
                 project_url=None,
                 template_dir=None,
                 examples_dir=None,
                 exclude=None):
        ...

    def generate(self, dest_dir):
        ...


     # ----------- those methods can be overriden ---------------

    def process_function_docstring(self, docstring, function):
        ...

    def process_class_docstring(self, docstring, class_):
        ...

    def process_method_docstring(self, docstring, method):
        ...

    def process_signature(self, signature):
        ...

```

Reply:
Looks great! I will review today. Thank you!